python-spec/src/somacore/coordinates.py

```diff
@@ -0,0 +1,491 @@
+            )
+
+    def __matmul__(self, other: Any) -> CoordinateTransform:
+        if self.input_axes != other.output_axes:
+            raise ValueError("Axis mismatch between transformations.")
+        if isinstance(other, IdentityTransform):
+            return AffineTransform(other.input_axes, self.output_axes, self._matrix)
+        if isinstance(other, AffineTransform):
+            return AffineTransform(
+                other.input_axes,
+                self.output_axes,
+                self.augmented_matrix @ other.augmented_matrix,
+            )
+        raise TypeError(
+            f"Cannot multiply a CoordinateTransform by type {type(other)!r}."
+        )
+
+    def __rmatmul__(self, other: Any) -> CoordinateTransform:
+        if other.input_axes != self.output_axes:
+            raise ValueError("Axis mismatch between transformations.")
+        if isinstance(other, IdentityTransform):
+            return AffineTransform(
+                self.input_axes, other.output_axes, self.augmented_matrix
+            )
+        if isinstance(other, AffineTransform):
+            return AffineTransform(
+                self.input_axes,
+                other.output_axes,
+                other.augmented_matrix @ self.augmented_matrix,
+            )
+        raise TypeError(
+            f"Cannot multiply a CoordinateTransform by type {type(other)!r}."
+        )
+
+    @property
+    def augmented_matrix(self) -> npt.NDArray[np.float64]:
+        """Returns the augmented affine matrix for the transformation.
+
+        Lifecycle: experimental
+        """
+        return self._matrix
+
+    def inverse_transform(self) -> CoordinateTransform:
+        """Returns the inverse coordinate transform.
+
+        Lifecycle: experimental
+        """
+        rank = len(self.output_axes)
+        inv_a = np.linalg.inv(self._matrix[:-1, :-1])
+        b2 = -inv_a @ self._matrix[:-1, -1].reshape((rank, 1))
+        inv_augmented: npt.NDArray[np.float64] = np.vstack(
+            (
+                np.hstack((inv_a, b2)),
+                np.hstack((np.zeros(rank), np.array([1]))),
+            )
+        )
+        return AffineTransform(self.output_axes, self.input_axes, inv_augmented)
+
+
+class ScaleTransform(AffineTransform):
+    """A scale coordinate transformation from one coordinate space to another.
+
+    Args:
+        input_axes: The names of the axes for the input coordinate space.
+        output_axes: The names of the axes for the output coordinate space.
+        scale_factors: The scale factors for the transformation. There must be one
+            value per axis.
+
+    Lifecycle: experimental
+    """
+
+    def __init__(
+        self,
+        input_axes: Union[str, Sequence[str]],
+        output_axes: Union[str, Sequence[str]],
+        scale_factors: npt.ArrayLike,
+    ):
+        super(AffineTransform, self).__init__(input_axes, output_axes)
```

Comment:
Would it be better to generate a matrix from the `scale_factors` to pass to the parent? Right now this ends up skipping over the `AffineTransform`’s `__init__`, which means it doesn’t fulfill the expectation that `_matrix` is set which other methods from `AffineTransform` would expect.

A few solutions:

- Pass through the expected inputs to each (so `UniformScaleTransform` calls `super().__init__(ins, outs, [scale_factor] * len(ins)`; `ScaleTransform` calls `super().__init__(ins, outs, scale_factors_matrix)`, etc. That will ensure that all the invariants of the entire class hierarchy are satisfied.

- Use an inheritance structure so all the shared stuff lives in `_AffineBase`, and each subclass contains only the behavior specific to its type:
  ```
  _AffineBase (abstract)
  +-- AffineTransform (has matrix)
  +-- ScaleTransform (has list)
  +-- UniformScaleTransform (has single element)
  ```

- Don’t special-case scalings and uniform scalings, and use a factory method:
  ```python
  class AffineTransform(...):
    def __init__(self, ins, outs, matrix):
      ...

    @classmethod
    def scaling(cls, ins, outs, factors) -> Self:
      ...
      return cls(ins, outs, factor_matrix)

    @classmethod
    def uniform_scaling(cls, ins, outs, factor) -> Self:
      ...
    ```